ui/src/App.js

```diff
@@ -121,10 +122,32 @@ class App extends Component {
     datasetApi.datasetGet(datasetCallback);
   }
 
-  handleChange(selectedOption) {
+  handleChange(selectedFacets) {
+    let deletedFacets = this.state.selectedFacets.filter(
+      x => !selectedFacets.includes(x)
+    );
+    deletedFacets.forEach(removed => {
+      if (this.filterMap.get(removed.label) !== undefined) {
+        this.filterMap.delete(removed.label);
+      }
+    });
+    let filterArray = this.filterMapToArray(this.filterMap);
+    this.setState({ filter: filterArray });
+
     let extraFacets = [];
-    selectedOption.forEach(option => extraFacets.push(option.value));
-    console.log(extraFacets);
+    selectedFacets.forEach(option => extraFacets.push(option.value));
+    if (extraFacets.length > 0) {
+      this.setState({ extraFacets: extraFacets });
+      this.setState({ selectedFacets: selectedFacets });
+      this.facetsApi.facetsGet(
+        { filter: filterArray, extraFacets: extraFacets },
+        this.facetsCallback
+      );
+    } else {
+      this.setState({ extraFacets: [] });
+      this.setState({ selectedFacets: [] });
+      this.facetsApi.facetsGet({ filter: filterArray }, this.facetsCallback);
```

Comment:
setState() triggers a render(). (See ["Thanks to the setState() call,"](https://reactjs.org/docs/state-and-lifecycle.html).) Let's call setState() only once. It's ok if an argument to setstate() is an empty list.

Similarly, let's have only one facetsGet() call, instead of lines 142 and 149. It's ok if extraFacets is empty.

Reply:
I think there will be only one setState() call per request. In an if-else branch, only if or else part will be called at a time.
I think I explicitly added the empty arrays because I got an error, I can check again.